### PR TITLE
Gadt conc

### PR DESCRIPTION
**Pull request by QHWU1228**
This branch improves the design of data_cons in syntax by introducing the universal and existential type variables.

### CODE REVIEW

File: src/G2/Execution/Rules.hs [hunk 1]
```diff
@@ -792,7 +790,8 @@ defAltExpr (_:xs) = defAltExpr xs
 
 -- | Creates and applies new symbolic variables for arguments of Data Constructor
 concretizeSym :: [(Id, Type)] -> Maybe Coercion -> (State t, NameGen) -> DataCon -> ((State t, NameGen), Expr)
-concretizeSym bi maybeC (s, ng) dc@(DataCon _ ts) =
```
Comment (BillHallahan) [inline src/G2/Execution/Rules.hs:795]: Let's mark this with "TODO", to make it easy to grep for

File: src/G2/Execution/Rules.hs [hunk 2]
```diff
@@ -1186,7 +1185,8 @@ retReplaceSymbFuncTemplate s@(State { expr_env = eenv
         (x, ng') = freshId t1 ng
         (x', ng'') = freshId t1 ng'
         (alts, symIds, ng''') =
-            foldr (\dc@(DataCon _ dcty) (as, sids, ng1) ->
```
Comment (BillHallahan) [inline src/G2/Execution/Rules.hs:1189]: Mark with TODO

File: src/G2/Language/Monad/AST.hs [hunk 3]
```diff
@@ -162,7 +162,12 @@ instance ASTContainerM DataCon Expr where
     modifyContainedASTsM _ dc = return dc
 
 instance ASTContainerM DataCon Type where
-    modifyContainedASTsM f (DataCon n t) = return . DataCon n =<< f t
+    modifyContainedASTsM f (DataCon n t u e) = do
+        t' <- f t 
+        u' <- traverse (modifyContainedASTsM f) u 
+        e' <- traverse (modifyContainedASTsM f) e
```
Comment (BillHallahan) [inline src/G2/Language/Monad/AST.hs:168]: Don't need to use traverse here- just `modifyContainedASTsM f u`, works, for instance, because of ASTContainerM instance for lists.

File: src/G2/Language/Naming.hs [hunk 4]
```diff
@@ -66,7 +67,7 @@ import qualified Data.Sequence as S
 import qualified Data.Text as T
 import Data.Tuple
 import qualified Text.Builder as TB
-
+import qualified Data.Monoid as M
```
Comment (BillHallahan) [inline src/G2/Language/Naming.hs:70]: Do we need this?  Don't see how we're using it?

File: src/G2/Language/Arbitrary.hs [hunk 5]
```diff
@@ -169,7 +169,8 @@ arbDataCon tenv unq ret_ty = do
     n <- chooseEnum ('A', 'Z')
     ar_c <- chooseInt (0, 5)
     ar_ty <- vectorOf ar_c (sized $ \k -> arbType tenv (k `div` ar_c))
-    return $ DataCon (Name (T.singleton n) Nothing unq Nothing) (mkTyFun $ ar_ty ++ [ret_ty])
```
[the remark this answers is not in the document]
Comment (QHWU1228) [reply src/G2/Language/Arbitrary.hs:172]: I was gonna discuss with you in the meeting about those. Also, I think I fix those errors already.

File: src/G2/Language/Naming.hs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE InstanceSigs #-}
```
Comment (BillHallahan) [inline src/G2/Language/Naming.hs:5]: Don't think we need this?

Comment (QHWU1228) [reply src/G2/Language/Naming.hs:5]: This is not the final version. I wass gonna push the final version after our meeting. I am not expecting you to check this right now.

File: src/G2/Language/Naming.hs [hunk 7]
```diff
@@ -337,8 +338,9 @@ renameVars' _ _ e = e
 renameExprId :: Name -> Name -> Id -> Id
 renameExprId old new (Id n t) = Id (rename old new n) t
 
+-- we should only rename the name in the DataCon and leave univ and exist type variable intact right?
```
Comment (BillHallahan) [inline src/G2/Language/Naming.hs:341]: Correct

File: src/G2/Language/Naming.hs [hunk 8]
```diff
@@ -374,8 +376,9 @@ renamesExprs' _ e = e
 renamesExprId :: HM.HashMap Name Name -> Id -> Id
 renamesExprId hm (Id n t) = Id (renames hm n) t
 
+-- only rename the name in the DataCon while leaving univ and exist type variable inact?
```
Comment (BillHallahan) [inline src/G2/Language/Naming.hs:379]: Correct

File: src/G2/Language/Naming.hs [hunk 9]
```diff
@@ -433,15 +436,21 @@ instance Named Alt where
 
 instance Named DataCon where
     {-# INLINE names #-}
-    names (DataCon n t) = n S.<| names t
+    names (DataCon n t u e) = 
+        let 
+            all_tyvars = names u S.>< names e
+            t_ns = names t 
+        in 
+        n S.<| t_ns S.>< all_tyvars
```
Comment (BillHallahan) [inline src/G2/Language/Naming.hs:444]: Do we really need a let for these bindings?  Nitpicky, but sometimes just writing things inline is clearer, especially if lots of similar operations (just calling names on a bunch of stuff)

File: src/G2/Language/Naming.hs [hunk 10]
```diff
@@ -179,7 +179,7 @@ altMatchNames (DataAlt dc i) = dataConName dc ++ (map idName i)
 altMatchNames _ = []
 
 dataConName :: DataCon -> [Name]
-dataConName (DataCon n _) = [n]
+dataConName (DataCon n _ u e) = [n] ++ map idName u ++ map idName e
```
Comment (BillHallahan) [inline src/G2/Language/Naming.hs:182]: We just want to return the data constructor name here, we do not want to return the type names (revert this change)

File: src/G2/Language/AST.hs [hunk 11]
```diff
@@ -309,8 +309,8 @@ instance ASTContainer DataCon Expr where
     modifyContainedASTs _ d = d
 
 instance ASTContainer DataCon Type where
-    containedASTs (DataCon _ t) = [t]
-    modifyContainedASTs f (DataCon n t) = DataCon n (f t)
+    containedASTs (DataCon _ t u e) = [t] ++ concatMap containedASTs u ++ concatMap containedASTs e
+    modifyContainedASTs f (DataCon n t u e) = DataCon n (f t) (map (modifyContainedASTs f) u) (map (modifyContainedASTs f) e)
```
Comment (BillHallahan) [inline src/G2/Language/AST.hs:313]: Don't need to use concatMap/map- just call containedASTs/modifyContainedASTs on the whole lists

File: src/G2/QuasiQuotes/Internals/G2Rep.hs [hunk 12]
```diff
@@ -146,9 +146,11 @@ qqDataConLookupFallBack tyv_n arg_n qqtn qqdc type_nm_qqm dc_nm_qqm tenv
             i = Id n TYPE
 
             t = mkTyFun $ replicate (arg_n + 1) (TyCon n TYPE)
-            t' = foldr TyForAll t (replicate tyv_n i)
+            ty_forall_ids = replicate tyv_n i
+            t' = foldr TyForAll t ty_forall_ids
+            -- might need a double check for this update
```
Comment (BillHallahan) [inline src/G2/QuasiQuotes/Internals/G2Rep.hs:151]: Looks good!

Comment (QHWU1228) [reply src/G2/QuasiQuotes/Internals/G2Rep.hs:151]: So I think those are all the changes I need to make before merging it into the master right?

File: src/G2/Language/Naming.hs (hunk 10 above)
Comment (BillHallahan) [reply src/G2/Language/Naming.hs:182]: That is, just want to return `n` (notice how this function is used- the only functions that call it are just supposed to return value level names, not type level names.)

Comment (QHWU1228) [reply src/G2/Language/Naming.hs:182]: I got it, sorry for my mistakes. I think because I was updating this along with the containedASTs which make me think we should get all the names from the function. I assume those three changes and changes I need to make before merging into the master right?
